### DISCUSSION TITLE
Is it currently possible to use AI-deck and Loco positioning together on a drone?

### DISCUSSION THREAD

**Discussion by ArtemRedchych** [bitcraze/discussions#563] [Q&A] [answered]
I am interested in using AI-deck and loco positioning on my drone, but I am seeing a message on my Loco positioning device that says "INFO:cfclient.utils.input.inputreaders.pysdl2:Found 0 devices". I saw on the Craziflie website that they mention support for this feature will be available in a future software update. Can anyone confirm whether or not it is currently possible to use AI-deck and loco positioning together, and if so, what steps I need to take to get this working? Thank you!

**Comment by tobbeanton** [comment, marked as answer]
The two decks have conflicting pins AI:TX1, RX1, NINA_SYSBOOT and Loco:IRQ, RST, CS. The Loco pins could be moved by the offered patching-pads but the board still conflicts with NINA_SYSBOOT and Loco:CS. The NINA_SYSBOOT on the other hand is only used during startup (for bootloading) so they might not conflict. I did a quick test by changing the AIdeck driver like so:

```
static const DeckDriver aideck_deck = {
    .vid = 0xBC,
    .pid = 0x12,
    .name = "bcAI",

    .usedGpio = /*DECK_USING_IO_1 | */ DECK_USING_IO_4,
    .usedPeriph = 0, // DECK_USING_UART2,

    .memoryDef = &espMemoryDef,
    .memoryDefSecondary = &gap8MemoryDef,

    .init = aideckInit,
    .test = aideckTest,
};
```

Then Lodo deck boots and GAP8 but not NINA (ESP32) so something else needs to be done. Perhaps Loco deck manages to put ESP32 into bootloader mode.

**Reply by ArtemRedchych** [reply]
Could you please provide guidance on where to locate the AIdeck driver file in the Crazyflie source code and where exactly to place the code you provided?

**Reply by tobbeanton** [reply]
It's in [aideck.c](https://github.com/bitcraze/crazyflie-firmware/blob/479464f2073a76b81830acf83b734847a7582b1a/src/deck/drivers/src/aideck.c#L373)

**Reply by ArtemRedchych** [reply]
I want to share an update with you. I followed the steps you provided, including cloning the Crazyflie firmware repository, modifying the aideck.c file, building the firmware, and uploading it to the drone using the instructions from https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/building-and-flashing/build/. I am happy to report that the Loco positioning is now working with the AI deck. Thank you very much for your help!

**Reply by joatjoses** [reply]
Hello @ArtemRedchych! I'm also trying to use both the AI deck and the loco positioning deck. Did you do other steps to make the two decks be compatible? I am experiencing problems making the loco positioning deck work when paired with the AI deck as seen in this [discussion](https://github.com/orgs/bitcraze/discussions/776). Thank you for you help.

**Comment by sajidj-sketch** [comment]
# Solved: AI-deck 1.1 + Loco deck fully coexisting (GAP8 + ESP32 WiFi + LPS positioning simultaneously)

**Setup:** Crazyflie 2.1, AI-deck 1.1, Loco deck, crazyflie-firmware `2025.02` (d89b0c7a).
**Result:** both decks enumerate and run at once — GAP8 app running, ESP32 up with WiFi (CPX intact), LPS ranging + position estimate valid. Unlike the earlier `.usedPeriph = 0` workaround in this thread, the ESP32 stays fully functional.

## 1. Pin conflict map (AI-deck 1.1 datasheet Rev 2, §3)

| Pin | AI-deck | Loco (stock) | Resolution |
|---|---|---|---|
| RX1/TX1 | UART1 → GAP8 | IRQ / RST (GS1/GS2) | cut GS1+GS2, reroute |
| IO1 | ESP32 **SYSBOOT** (strap, sampled at ESP32 reset release) | SPI CS (fixed, not routable) | share (safe — see §2) |
| IO2 | — | alt IRQ | use for IRQ |
| IO3 | — | alt RST | use for RST |
| IO4 | **reset to both GAP8 and ESP32** | — | never route Loco RST here |
| RX2/TX2 | UART2 → ESP32 (CPX/WiFi) | — | keep UART2 in firmware |

## 2. Why Loco RST on IO4 kills the ESP32 (deterministic, not a race)

`locodeck.c` init does, in this order:

```c
pinMode(CS_PIN, OUTPUT);          // CS = IO1: STM32 output register defaults to 0
                                  // → IO1 drives LOW; first SPI txn (raising CS) is later
digitalWrite(GPIO_PIN_RESET, 0);  // 10 ms reset pulse
vTaskDelay(M2T(10));
digitalWrite(GPIO_PIN_RESET, 1);  // release → ESP32 samples SYSBOOT (= IO1 = LOW) NOW
```

If RST is on IO4, the pulse resets GAP8+ESP32; the ESP32 leaves reset with SYSBOOT low → enters **ROM serial bootloader** → app never runs. Symptom: GAP8 LED blinks, ESP32 LED dead, no WiFi; remove the LPS deck and the ESP32 is fine. GAP8 survives because its boot straps are not on IO1. With RST on IO3 the Loco never resets the ESP32, and IO1 sharing is harmless (SYSBOOT only sampled at reset release).

## 3. Hardware rework (Loco deck only; AI-deck untouched)

1. **Cut GS1, GS2** (frees RX1/TX1 → UART1 for GAP8)
2. **Bridge GS3** (IRQ → IO2)
3. **Wire DWM1000 RST → IO3** (not IO4)

Final: CS=IO1, IRQ=IO2, RST=IO3.

Continuity checks, power off (IO3/IO4 are adjacent holes — a stray bridge here reproduces the IO4 failure exactly; it bit me):
RST↔IO3 beeps · RST↔IO4 **silent** · IO3↔IO4 **silent** · IRQ↔IO2 beeps · CS↔IO1 beeps.

## 4. Firmware changes (3 items)

**(a) Kconfig** (`make menuconfig` → Expansion deck configuration):

```
CONFIG_DECK_LOCODECK_USE_ALT_PINS=y        # IRQ=IO2, RST=IO3
# CONFIG_LOCODECK_ALT_PIN_RESET is not set # must stay OFF — ON selects IO4
```

plus your WiFi setup (`CONFIG_DECK_AI_WIFI_SETUP_STA=y` or `_AP=y`, `CONFIG_DECK_AI_SSID`, `CONFIG_DECK_AI_PASSWORD`). ⚠ `make cf2_defconfig` silently resets all of these.

**(b) `src/deck/drivers/src/aideck.c`** — release IO1 to the Loco (resource manager otherwise refuses to init the second deck; both drivers declare IO1). Keep IO4 and UART2:

```diff
-    .usedGpio = DECK_USING_IO_1 | DECK_USING_IO_4,
+    .usedGpio = DECK_USING_IO_4,
     .usedPeriph = DECK_USING_UART2,   // keep! UART2 carries CPX/WiFi; 0 kills the ESP32 link
```

Safe because the driver only *toggles* IO1/IO4 during deliberate ESP32 re-flashing over CPX (`resetEspToBootloader()`), not in normal operation.

**(c) `src/deck/drivers/src/locodeck.c`** — make CS idle high *before* the DW1000 reset pulse. Correct SPI hygiene, and closes the §2 trap even if the reset lines are ever tied together again (design or solder bridge):

```diff
   // Init pins
+  digitalWrite(CS_PIN, 1);   // pre-load ODR: pin comes up driving HIGH, no low glitch
   pinMode(CS_PIN, OUTPUT);
   pinMode(GPIO_PIN_RESET, OUTPUT);
   pinMode(GPIO_PIN_IRQ, INPUT);
```

## 5. Build & flash

```bash
make menuconfig          # options from 4(a)
make -j$(nproc)          # → build/cf2.bin
make cload               # or: make cload CLOAD_CMDS="-w radio://0/80/2M"
```

## 6. Verify

- Console: both decks init, no resource conflict/assert
- GAP8 LED blinks; ESP32 LED blinks, WiFi up (I stream from a custom ESP32 UDP firmware)
- LPS ranging active, valid position estimate

## 7. Caveats

- **Don't re-flash ESP32/GAP8 over CPX while the Loco is active** — that path drives IO1 (SYSBOOT) and IO4, which now interact with the running Loco. Detach the LPS deck for AI-deck firmware updates.
- IO1 remains physically shared (CS vs SYSBOOT); with fix 4(c) the boot window is closed and runtime CS traffic is harmless.
- Only IO2/IO3 are free alongside the AI-deck — this rework consumes both, so it won't stack with decks needing IO2/IO3 (e.g. flow deck's IO3) without further patching.
